curve25519-dalek/src/encodable_curve.rs

```diff
@@ -0,0 +1,199 @@
+//
+// This file is part of curve25519-dalek fork by Arcium.
+//! Module for common traits.
+
+use core::ops::ShrAssign;
+
+use elliptic_curve::array::Array;
+use elliptic_curve::bigint::ArrayEncoding;
+use elliptic_curve::bigint::U256;
+use elliptic_curve::consts::U32;
+use elliptic_curve::ops::Invert;
+use elliptic_curve::ops::MulByGenerator;
+use elliptic_curve::scalar::FromUintUnchecked;
+use elliptic_curve::scalar::IsHigh;
+use elliptic_curve::Curve;
+use elliptic_curve::Field;
+use elliptic_curve::FieldBytes;
+use elliptic_curve::FieldBytesEncoding;
+use elliptic_curve::PrimeField;
+use elliptic_curve::ScalarPrimitive;
+use subtle::Choice;
+use subtle::CtOption;
+
+use crate::constants::BASEPOINT_ORDER_PRIVATE;
+use crate::EdwardsPoint;
+use crate::Scalar;
+
+// Empty struct for the curve
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord)]
+pub struct Dalek {}
+
+impl Curve for Dalek {
+    type FieldBytesSize = U32;
+    type Uint = U256;
+    const ORDER: Self::Uint = U256::from_le_slice(BASEPOINT_ORDER_PRIVATE.as_bytes());
+}
+
+// Impls for EdwardsPoint
+impl MulByGenerator for EdwardsPoint {
+    fn mul_by_generator(scalar: &Self::Scalar) -> Self {
+        <Self as group::Group>::generator() * scalar
+    }
+}
+
+// Impls for Scalar
+impl AsRef<Scalar> for Scalar {
+    fn as_ref(&self) -> &Scalar {
+        self
+    }
+}
+
+// NOTE: ScalarPrimitive is big endian. Dalek uses little endian
+impl From<ScalarPrimitive<Dalek>> for Scalar {
+    fn from(value: ScalarPrimitive<Dalek>) -> Self {
+        Scalar::from_bytes_mod_order(value.to_uint().to_le_bytes())
+    }
+}
+
+impl From<U256> for Scalar {
+    fn from(value: U256) -> Self {
+        Scalar::from_bytes_mod_order(value.to_le_bytes())
+    }
+}
+
+impl FromUintUnchecked for Scalar {
+    type Uint = U256;
+    fn from_uint_unchecked(uint: Self::Uint) -> Self {
+        Scalar {
+            bytes: uint.to_le_bytes(),
+        }
+    }
+}
+
+impl Into<FieldBytes<Dalek>> for Scalar {
+    fn into(self) -> FieldBytes<Dalek> {
+        U256::encode_field_bytes(&U256::from_le_slice(&self.bytes))
+    }
+}
+
+impl Into<ScalarPrimitive<Dalek>> for Scalar {
+    fn into(self) -> ScalarPrimitive<Dalek> {
+        ScalarPrimitive::new(U256::from_le_slice(&self.bytes)).unwrap()
+    }
+}
+
+impl Into<U256> for Scalar {
+    fn into(self) -> U256 {
+        U256::from_le_slice(&self.bytes)
+    }
+}
+
+impl Into<U256> for &mut Scalar {
+    fn into(self) -> U256 {
+        U256::from_le_slice(&self.bytes)
+    }
+}
+
+impl Invert for Scalar {
+    type Output = CtOption<Self>;
+    fn invert(&self) -> Self::Output {
+        Field::invert(self)
+    }
+}
+
+impl IsHigh for Scalar {
+    fn is_high(&self) -> Choice {
+        let res = self >= &<Scalar as PrimeField>::TWO_INV;
+        Choice::from(res as u8)
+    }
+}
+
+impl PartialOrd for Scalar {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
```

Comment:
should we perhaps implement a constant-time version of this?

Reply:
I made it constant time and implemented Ord as well

Reply:
L125 implementation is not constant time

Reply:
He made it constant time then unmade it constant time.

Reply:
According to [this](https://www.rfc-editor.org/rfc/rfc9380#name-utility-functions), "all arithmetic, comparisons, and assignments MUST also be implemented in constant time", so let's stick to a constant time implementation. Wdyt @ibarrond ?

Reply:
Indeed, my last commit makes it NOT constant time. If anyone thinks that there is an advantage to having it be constant time, please speak up